// @skill-wallet/auth/build/components/BackButton.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx } from "react/jsx-runtime";
import { ArrowBackIos } from '@mui/icons-material';
import { IconButton } from '@mui/material';
var BackButton = function (_a) {
    var handleClick = _a.handleClick;
    var handleOnClick = function () {
        handleClick();
    };
    return (_jsx(IconButton, __assign({ onClick: handleOnClick, sx: {
            borderRadius: '30px',
            '&:hover': {
                background: '#FFFFFF',
                '& .backIcon': {
                    color: '#000000',
                },
            },
            width: '45px',
            height: '45px',
        } }, { children: _jsx(ArrowBackIos, { color: "secondary", className: "backIcon", sx: {
                pl: '8px',
            } }, void 0) }), void 0));
};
export default BackButton;
