// @skill-wallet/auth/build/components/ErrorBox.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import { Typography } from '@mui/material';
import { Box } from '@mui/system';
import { SwButton } from 'sw-web-shared';
var ErrorBox = function (_a) {
    var errorData = _a.errorData;
    var handleOnClick = function () {
        console.log(errorData);
        if (errorData) {
            errorData.action();
        }
    };
    return (_jsxs(Box, __assign({ sx: {
            width: '100%',
            minHeight: '460px',
            display: 'flex',
            justifyContent: 'center',
            flexDirection: 'column',
            alignItems: 'center',
        } }, { children: [_jsx(Typography, __assign({ sx: { mb: '26px', width: '90%' }, variant: "h2" }, { children: errorData.errorMessage }), void 0), _jsx(SwButton, { sx: {
                    borderColor: 'primary.main',
                    mb: '26px',
                }, onClick: handleOnClick, btnType: "large", mode: "dark", label: errorData.actionLabel }, void 0)] }), void 0));
};
export default ErrorBox;
